File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int length = s.length();
        bool dp[length][length];
        memset(dp, 0, sizeof(dp));
        int max_pal_length = 1;
        for (int i = 0; i < length; i++){
            dp[i][i] = true;
        }
        int start = 0;
        for (int j = 0; j < length - 1; j++){
            if (s[j] == s[j + 1]){
                dp[j][j + 1] = true;
                start = j;
                max_pal_length = 2;
            }
        }
        for (int k = 3; k <= length; ++k) {
            for (int i = 0; i < length - k + 1; ++i) {
                int j = i + k - 1;
                if (dp[i + 1][j - 1] && s[i] == s[j]) {
                    dp[i][j] = true;
                    if (k > max_pal_length) {
                        start = i;
                        max_pal_length = k;
                    }
                }
            }
        }
        return s.substr(start, max_pal_length);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cs
public class Solution {
    public static int checkPalidrome(string s, int left, int right) {
        while (left>=0 && right<s.Length && s[left] == s[right]) {
            left--; right++;
        }
        return right - left - 1;
    }
    public string LongestPalindrome(string s) {
        if (s == "") return "";
        int start = 0;
        int end = 0;
        for (int ind = 0; ind < s.Length; ind++) {
            int odd_len = checkPalidrome(s, ind, ind);
            int even_len = checkPalidrome(s, ind, ind + 1);
            int len = odd_len > even_len ? odd_len : even_len;
            if (len > end - start) {
                start = ind - (len - 1) / 2;
                end = ind + len / 2;
            }
        }
        return s.Substring(start, end - start + 1);
    }
}